csharp/ql/src/semmle/code/csharp/PrintAst.qll

```diff
@@ -110,9 +110,16 @@ private predicate isCompilerGeneratedAttributable(Attributable attributable) {
 }
 
 /**
- * Retrieves the QL class(es) for entity `el`
+ * Retrieves the canonical QL class(es) for entity `el`
  */
-private string getQlClass(Element el) { result = "[" + concat(el.getAQlClass(), ",") + "] " }
+private string getQlClass(Element el) {
+  if count(el.getAPrimaryQlClass()) = 1 and not el.getAPrimaryQlClass() = "???"
+  then result = "[" + el.getAPrimaryQlClass() + "] "
+  else
+    if count(el.getAPrimaryQlClass()) > 1
+    then result = "[" + concat(el.getAPrimaryQlClass(), ",") + "] "
+    else result = "ERR [" + concat(el.getAQlClass(), ",") + "] "
+}
```

Comment:
This will force computation of `getAQlClass()`, which is often very expensive. I therefore think we should do the same as C++, i.e.

```
/**
 * Retrieves the canonical QL class(es) for entity `el`
 */
private string qlClass(ElementBase el) {
  result = "[" + concat(el.getAPrimaryQlClass(), ",") + "] "
  // Alternative implementation -- do not delete. It is useful for QL class discovery.
  //result = "["+ concat(el.getAQlClass(), ",") + "] "
}
```